app/abci.go

```diff
@@ -0,0 +1,161 @@
+package app
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lazyledger/lazyledger-app/x/lazyledgerapp/types"
+	abci "github.com/lazyledger/lazyledger-core/abci/types"
+	core "github.com/lazyledger/lazyledger-core/proto/tendermint/types"
+)
+
+const (
+	// runTxModeDeliver determines if the tx being ran should only be checked,
+	// or if it should actually modify the state.
+	// copy and pasted from the sdk, as it is not exported
+	// https://github.com/lazyledger/cosmos-sdk/blob/3addd7f65d1c35f76716d67641ee629b01f5b9c7/baseapp/baseapp.go#L28
+	runTxModeDeliver = 3
+)
+
+// This file should contain all of the altered ABCI methods
+
+// PreprocessTxs fullfills the lazyledger-core version of the ACBI interface, by
+// performing basic validation for the incoming txs, and by cleanly separating
+// share messages from transactions
+// todo(evan): refactor out a for loop.
+func (app *App) PreprocessTxs(txs abci.RequestPreprocessTxs) abci.ResponsePreprocessTxs {
+	squareSize := app.SquareSize()
+	shareCounter := uint64(0)
+	var shareMsgs []*core.Message
+	var processedTxs [][]byte
+	for _, rawTx := range txs.Txs {
+		// decode the Tx
+		tx, err := app.txConfig.TxDecoder()(rawTx)
+		if err != nil {
+			continue
+		}
+
+		// don't process the tx if the transaction doesn't contain a
+		// PayForMessage sdk.Msg
+		if !hasWirePayForMessage(tx) {
+			processedTxs = append(processedTxs, rawTx)
+			continue
+		}
+
+		// only support transactions that contain a single sdk.Msg
+		if len(tx.GetMsgs()) != 1 {
+			continue
+		}
+
+		msg := tx.GetMsgs()[0]
+
+		// run basic validation on the transaction
+		err = tx.ValidateBasic()
+		if err != nil {
+			continue
+		}
+
+		// process the message
+		coreMsg, signedTx, err := app.processMsg(msg)
+		if err != nil {
+			continue
+		}
+
+		// execute the tx in runTxModeDeliver mode (3)
+		// execution includes all validation checks burning fees
+		// currently, no fees are burned
+		_, _, err = app.BaseApp.TxRunner()(runTxModeDeliver, rawTx)
+		if err != nil {
+			continue
+		}
```

Comment:
Actually, didn't we say to not execute the Tx immediately until we add explicit ABCI methods for that purpose?
See: https://github.com/tendermint/spec/issues/208#issuecomment-721158558
and https://github.com/tendermint/spec/pull/254

Also, if run the Tx during preprocess, won't that app still try to execute the Tx from the previous block during, too?

Reply:
The sdk uses the [`runTx`](https://github.com/lazyledger/cosmos-sdk/blob/3addd7f65d1c35f76716d67641ee629b01f5b9c7/baseapp/baseapp.go#L560) method to run stateful checks and execute any state changing transactions. `runTx` provides those transactions with access to the state via the `anteHandler`s. While the `runTxModeDeliver` flag is being passed to `runTx`, 

> Also, if run the Tx during preprocess, won't that app still try to execute the Tx from the previous block during, too?

Under the hood, there are two distinct transactions (what the sdk refers to as `sdk.Msgs`). This is where the naming gets a tad confusing. There's [`SignedTransactionDataPayForMessage`](https://github.com/lazyledger/lazyledger-app/blob/a41e189f754ae31c95dbcf6edee6a2626f321563/x/lazyledgerapp/types/tx.pb.go#L288), which is not being pushed through `runTx` and is instead returned by `PreprocessTxs`, and there is also the [`MsgWirePayForMessage`](https://github.com/lazyledger/lazyledger-app/blob/a41e189f754ae31c95dbcf6edee6a2626f321563/x/lazyledgerapp/types/tx.pb.go#L34), which is passed to `PreprocessTxs` and is being pushed through `runTx`. Each is ran separately, and each has its own signature. 

We can look at what the handler for `MsgWirePayForMessage` does [here](https://github.com/lazyledger/lazyledger-app/blob/a41e189f754ae31c95dbcf6edee6a2626f321563/x/lazyledgerapp/keeper/keeper.go#L42), which is nothing atm. There is not currently a handler for `SignedTransactionDataPayForMessage`, which will throw an error if it is ran later via `DeliverTx`. While throwing an error if `SignedTransactionDataPayForMessage` is passed to `DeliverTx` was intentional, I think having a handler that for it that always throws an error would be more explicit.

This brings up that I should have better documentation for all of this "under-the-hood" stuff, even if some of it is temporary. Where would the best place for that be? I was planning on making a full blown integration test later #24, but this might need to be done in this PR.

Reply:
Yeah, the `runTxModeDeliver` sounds like it will actually deliver the Txs. That is what confused me. Thanks for clarifying.

> This brings up that I should have better documentation for all of this "under-the-hood" stuff, even if some of it is temporary. Where would the best place for that be?

This brings us back to that ADR discussion. Maybe we should document this kind of decision in light-weight ADRs that either precede or accompany PRs. And yes, (integration) tests would be a great way to increase confidence that everything behaves as we expect.

Reply:
After talking with @evan-forbes in person, we've decided to drop the runTx call and rely soley on checkTx for now. We will still have to take care of this in the future but we can likely tackle this together with immediate execution (which, as far as I understand is required for fee burning anyway).